Spark: Add builder pattern to SparkAppenderFactory to handle increasing number of arguments

szehon-ho opened the pull request:
As discussed in : https://github.com/apache/iceberg/pull/2362, rebased and submit part of my PR: https://github.com/apache/iceberg/pull/2240 about the SparkAppenderFactoryBuilder

szehon-ho commented:
Hi @aokolnychyi @RussellSpitzer , change is ready if you guys have some time to take a look, thanks

RussellSpitzer reviewed (commented):
This looks fine to me, but all the use cases replaced here don't seem like they benefit from the change, they all seem just as long? I'm probably just missing the spot in the original PR though.

szehon-ho commented:
Hi Russell, yea its just a code cleanup.

Both the aforementioned PR's (#2362 and #2240) noticed there's 4 versions of the SparkAppenderFactory  Constructor and that it becomes harder and harder to add new things, and can be cleaner like this.

I noticed that I added an extraneous sort-order pointer as part of copy and paste from my PR (#2240), and removed it as its now redundant with the table added by the #2362, updated the PR with the small fix

openinx commented on spark/src/main/java/org/apache/iceberg/spark/source/SparkAppenderFactory.java:
```diff
@@ -85,6 +72,52 @@
+    }
+
+    public SparkAppenderFactory build() {
+      return new SparkAppenderFactory(properties, writeSchema, dsSchema, spec, equalityFieldIds,
```
We will need to do the necessary validation before we construct the `SparkAppenderFactory` instance: 

1.  The `properties`, `writeSchema`, `dsSchema`  must not be null;
2.  If the `equalityFieldIds` is not null,  means people plan to write equality deletes into delete files. Then the `eqDeleteRowSchema` must not be null;  Another side,  if `eqDeleteRowSchema` is not null, then `equalityFieldIds` must not be null.
3.  the `posDeleteRowSchema` can always choose to be null or not.

szehon-ho replied:
Done

openinx commented on spark/src/main/java/org/apache/iceberg/spark/source/SparkAppenderFactory.java:
```diff
@@ -85,6 +72,52 @@
+    return new SparkAppenderFactoryBuilder(table, writeSchema, dsSchema);
+  }
+
+  public static class SparkAppenderFactoryBuilder {
```
Since the class `SparkAppenderFactory`  was not marked as `public`, so I think the newly introduced `SparkAppenderFactoryBuilder` also don't have to be `public` ? 

btw,   I think we could just use the name `Builder` because it's in the class `SparkAppenderFactory`,  the prefix `SparkAppenderFactory` seems to be redundant.

szehon-ho replied:
Done

szehon-ho commented:
Thanks for taking a look and the clarification of what should not be null @openinx , i made the suggested changes if you want to take another look.

openinx commented on spark/src/main/java/org/apache/iceberg/spark/source/SparkAppenderFactory.java:
```diff
@@ -85,6 +72,65 @@
     this.posDeleteRowSchema = posDeleteRowSchema;
   }
 
+  public static Builder builderFor(Table table, Schema writeSchema, StructType dsSchema) {
```
Nit: As iceberg is a library which will be embedded in user's project,  so we usually strictly control methods/classes marked as `public`.  In this case,  we don't have to mark this as `public`.  Similar comment in the following `Builder`'s  methods.

szehon-ho replied:
Done

openinx reviewed (commented):
Looks good to me overall,  just left a minor comment !

szehon-ho commented:
@openinx thanks again, yea I missed it

openinx commented:
Got this merged, thanks @szehon-ho for contributing.
